libcst/metadata/full_repo_manager.py

```diff
@@ -0,0 +1,107 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+# pyre-strict
+
+
+import json
+import subprocess
+from pathlib import Path
+from typing import (
+    TYPE_CHECKING,
+    Callable,
+    Dict,
+    List,
+    Mapping,
+    Optional,
+    Sequence,
+    Tuple,
+)
+
+from mypy_extensions import TypedDict
+
+import libcst as cst
+from libcst.metadata.type_inference_provider import (
+    InferredType,
+    PyreData,
+    TypeInferenceProvider,
+)
+from libcst.metadata.wrapper import MetadataWrapper
+
+
+if TYPE_CHECKING:
+    from libcst.metadata.base_provider import ProviderT
+
+
+class FullRepoManager:
+    def __init__(
+        self,
+        repo_root_dir: str,
+        paths: List[str],
+        providers: List["ProviderT"],
+        timeout: int = 5,
+    ) -> None:
+        self.root_path: Path = Path(repo_root_dir)
+        self._cache: Dict["ProviderT", Mapping[str, object]] = {}
+        self._timeout = timeout
+        for provider in providers:
+            handler = self._get_cache_handler(provider)
+            if handler:
+                self._cache[provider] = handler(paths)
```

Comment:
Correct me if I'm wrong, but this looks like it's making shell calls in the constructor.  IMO, it's best to do expensive processing in another method.  Constructing objects should be light

Reply:
I understand this concern, we need to run the batched pyre query before user calling.
If we want to make the constructor cheap, we can add one resolve step and cache the result.
```
m = FullRepoManager(root, paths, providers)
m.get_metadata_wrapper_for_path(path)  # it implicitly resolve the cache and store for reuse.
```

Reply:
The implicit resolve step is added as `_resolve_cache()`